app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+        }
+        require(DirectionHelper.KEY_CODES.contains(event.keyCode)) {
+            "Invalid key event passed to CursorController#directionKeyPress: $event"
+        }
+
+        val direction = when (event.keyCode) {
+            KeyEvent.KEYCODE_DPAD_UP -> Direction.UP
+            KeyEvent.KEYCODE_DPAD_DOWN -> Direction.DOWN
+            KeyEvent.KEYCODE_DPAD_LEFT -> Direction.LEFT
+            KeyEvent.KEYCODE_DPAD_RIGHT -> Direction.RIGHT
+            else -> return false
+        }
+        when (event.action) {
+            KeyEvent.ACTION_UP -> directionKeysPressed -= direction
+            KeyEvent.ACTION_DOWN -> {
+                directionKeysPressed += direction
+                pushCursorEvent(direction)
+            }
+            else -> return false
+        }
+
+        _isCursorMoving.onNext(!directionKeysPressed.isEmpty())
+
+        return true
+    }
+
+    private fun pushCursorEvent(direction: Direction) {
+        val edgeNearCursor = getEdgeOfScreenNearCursor()
+        val couldScroll = edgeNearCursor?.let { webViewCouldScrollInDirectionProvider(it) }
+
+        val cursorMovedToEdgeOfScreen = edgeNearCursor == direction
+        val endOfDomContentReached = couldScroll == false
+
+        val event = if (cursorMovedToEdgeOfScreen && endOfDomContentReached) {
+            CursorEvent.ScrolledToEdge(direction)
+        } else {
+            CursorEvent.CursorMoved(direction)
+        }
+
+        _cursorMovedEvents.onNext(event)
+    }
+
+    /**
+     * @return returns a MotionEvent if the event is consumed, null otherwise
+     */
+    @SuppressLint("Recycle")
+    @CheckResult(suggest = "Recycle MotionEvent after use")
+    private fun maybeToMotionEvent(event: KeyEvent): MotionEvent? {
+        if (!isCursorActive.blockingFirst()) {
+            return null
+        }
+        if (event.keyCode != KeyEvent.KEYCODE_DPAD_CENTER && event.keyCode != KeyEvent.KEYCODE_ENTER) {
+            return null
+        }
+
+        val motionEvent = MotionEvent.obtain(event.downTime, event.eventTime, event.action,
+                lastKnownCursorPos.x, lastKnownCursorPos.y, 0)
+        return when (event.action) {
+            KeyEvent.ACTION_UP -> {
+                _isSelectPressed.onNext(false)
+                motionEvent
+            }
+            KeyEvent.ACTION_DOWN -> {
+                _isSelectPressed.onNext(true)
+                motionEvent
+            }
+            else -> null
+        }
+    }
+
+    /**
+     * Called by CursorView in order to drive cursor movement.
+     *
+     * Performance:
+     * Previous code updated cursor VM position every 16MS (1_000 MS / 60 frames == 16.66). This
+     * led to minor jitter because we were not tied exactly to the actual framerate. This new
+     * implementation is harder to understand, but is noticeably more performant.
+     *
+     * Architecture:
+     * Having a View#onDraw call VM updates does not mesh well with our architecture, but we found
```

Comment:
nit: it's not clear from the docs that our draw/update loop is coupled and driven by `View.onDraw` (the line I'm commenting on mentions `onDraw` but doesn't really provide context) and that we're performing the cursor position update here.

Similarly, the kdoc on this class does not convey the architecture between these two classes (View & Model, how they interact and that we should look out for this method). I think we can do a better job being explicit about our touch points (class kdoc, this method, View kdoc). Sorry this isn't entirely clear but it's the end of the day and it'd be hard to convey this through comments without writing it myself. We could probably pair program to write out what is important, if you'd like.